商品購入機能

AG0811 opened the pull request:
＃What
PAYJPを導入・card.js作成
order群作成・adress群作成
order_adressモデル作成
素材よりorder.html.erb追加
soul out機能の実装
購入機能の単体テスト実装

＃Why
商品購入機能を実装するため


◆必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/5a82a8c840b9b38722929fa4957cd56a
◆入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/f0ace129a00a6628a638af8256db14a4
◆ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/b5effd005d9a33ab0867003b47dba3a5
◆ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/c46f6b39ea1240209942533a59dbf7c4
◆ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/bb0f733e8e21a4c0b38714848fc2bbf2
◆売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://gyazo.com/89defed98a2ba4902da230895d41233a
◆ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/89defed98a2ba4902da230895d41233a
◆ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/819698e9264c22c3e0d87600093a0564
◆テスト結果の画像
https://gyazo.com/1b28aac3bf05ff510fea844760de145f

AG0811 committed:
PAYJP追加１段階目

Issue
◆PAYPJのJSがhttp://localhost:3000/items/1/orders以外でも読み込まれているせいで、http://localhost:3000/items/1/ordersではリダイレクトしないと反映されない
◆リダイレクト時に入力していた内容が反映されていない

AG0811 committed:
一通り購入機能の実装とsoul outの実装

AG0811 committed:
単体テストを含め実装完了

Akinori0123 commented on app/javascript/card.js:
```diff
@@ -0,0 +1,44 @@
+  const payjp = Payjp(publicKey) // PAY.JPテスト公開鍵
+
+  const elements = payjp.elements();
+  console.log("elementsが呼び出されました");
```
```
console.log("elementsが呼び出されました");
```
デバッグ用のconsole.logは実装完了後削除しましょう。
同じファイルにあるconsole.logを全て削除いただければと思います。

Akinori0123 commented on app/views/items/index.html.erb:
```diff
@@ -136,9 +136,9 @@
             <div class='sold-out'><span>Sold Out!!</span></div>
-            <%# <% end %>
+            <% end %>
             <%# //商品が売れていればsold outを表示しましょう %>
```
```
<%# //商品が売れていればsold outを表示しましょう %>
```
138,142行目の実装案内コメントは内容を実装済みのため削除しましょう。

Akinori0123 commented on app/views/items/show.html.erb:
```diff
@@ -9,9 +9,9 @@
+      <% if Order.exists?(item_id: @item.id) %>
+      <div class="sold-out"><span>Sold Out!!</span></div>
+      <% end %>
       <%# //商品が売れている場合は、sold outを表示しましょう %>
```
```
<%# //商品が売れている場合は、sold outを表示しましょう %>
```
11,15行目の実装案内コメントも削除しましょう。

Akinori0123 commented on app/views/orders/index.html.erb:
```diff
@@ -0,0 +1,132 @@
+    </h1>
+    <%# 購入内容の表示 %>
+    <div class='buy-item-info'>
+      <%= image_tag "item-sample.png", class: 'buy-item-img' %>
```
```
image_tag "item-sample.png"
```
上記image_tagにこれから購入する商品の画像が表示されるようにしましょう。

Akinori0123 commented on app/views/orders/index.html.erb:
```diff
@@ -0,0 +1,132 @@
+        </h2>
+        <div class='buy-item-price'>
+          <p class='item-price-text'>¥<%= @item.price %></p>
+          <p class='item-price-sub-text'><%= '配送料負担' %></p>
```
```
<%= '配送料負担' %>
```
上記にこれから購入する商品の配送料負担情報が表示されるようにしましょう。

Akinori0123 commented on spec/factories/order_address.rb:
```diff
@@ -0,0 +1,15 @@
+    building_name { 'Test Building' }
+    phone_number { '09012345678' }
+    user_id { 1 }
+    item_id { 1 }
```
```
    user_id { 1 }
    item_id { 1 }
```
上記ですと存在しない商品やユーザーのidを指定することになるため、上記は削除しましょう。
（商品やユーザーを用意する方針について後のコメントで共有させていただきます）

Akinori0123 commented on spec/factories/order_address.rb:
```diff
@@ -0,0 +1,15 @@
+    item_id { 1 }
+    price { 1000 }
+    token { 'tok_abcdefghijk00000000000000000' }
+    item { association :item }
```
```
item { association :item }
```
フォームオブジェクトであるorder_addressはアソシエーションを組めないため、上記も削除しましょう。

Akinori0123 commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,84 @@
+
+RSpec.describe OrderAddress, type: :model do
+  before do
+    @order_address = FactoryBot.build(:order_address)
```
factorybotでは商品や購入者を用意することができないため、before内において商品や購入者を用意するようにしましょう。
以下の雛形や、またFormオブジェクトの参考カリキュラムをご確認いただければと思います。
```
before do
  item = <Factorybotを用いてitemを作成>
  user = <Factorybotを用いてuserを作成>
  @order_address = FactoryBot.build(:order_address, item_id: <itemのid> ,user_id: <userのid>)
end
```

Akinori0123 commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,84 @@
+    context '有効な属性の場合' do
+      it 'すべての属性が有効であること' do
+        expect(@order_address).to be_valid
+      end
```
・正常系テスト
要件を満たしていることを確認するために、建物名が空でも保存できることを確認しましょう。

Akinori0123 commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,84 @@
+        @order_address.phone_number = '090123'
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include("Phone number is too short")
+      end
```
・異常系テスト
電話番号が短い時のテストの実装良いですね。
要件を満たしていることを確認するために、電話番号が長い場合も無効であることをテストしましょう。

AG0811 replied:
レビューありがとうございます。
修正次第、再度レビュー依頼をさせていただきます。

Akinori0123 reviewed (commented):
レビュー依頼いただきありがとうございます！
コミット内容確認いたしました。コメントいたしましたのでご確認のほどよろしくお願いいたします。

AG0811 committed:
レビュー対応

＃What
購入確認画面での購入確認画面の画像と配送料負担情報の修正
実装案内コメントの消去
デバッグ用のconsole.log消去
購入機能の単体テスト加筆・修正

＃Why
商品購入機能を実装するため

◆【再】テスト結果の画像
https://gyazo.com/57682515c312a781956178ff759bc8ba

wad-sho commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,96 @@
+        @order_address.phone_number = '090-1234-5678'
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include("Phone number is invalid. Input only numbers")
+      end
```
以下テストを、異常系テストに追記しておきましょう！

```
・user_idが空では登録できないこと
・item_idが空では登録できないこと
```

AG0811 replied:
レビューありがとうございます。
対応次第再度レビュー依頼させていただきます。

wad-sho reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。

コメントいたしましたので、ご確認頂けますと幸いです。
修正の後、再度レビューのご依頼お願い致します！

AG0811 committed:
レビュー対応【再】

＃What
購入機能の単体テスト加筆修正
userとitemのidがnilの場合追加

＃Why
商品購入機能を実装するため
レビュー対応のため

Akinori0123 reviewed (commented):
レビュー依頼いただきありがとうございます！
コミット内容確認いたしました。LGTMです。
マージ後、mainブランチでpullしていただければと思います。

実装お疲れ様でした。
引き続き学習頑張っていきましょう！

AG0811 commented:
レビューありがとうございます。
LGTMを確認いたしましたのでmergeします。
ご対応ありがとうございます。
